### ISSUE TITLE
Flickering Issue on 5700XT

### ISSUE THREAD

**Issue by Zoomao** [panda3d/panda3d#981]
## Description

There is an issue in which textures and words flicker when the installed graphics card is the 5700XT. According to other people it seems to be an issue that specifically only affects people with the 5700XT. 
## Steps to Reproduce

While it can only be reproduced on a 5700XT, the flickering is evident in two of the samples. First, in the advanced cartoon shader sample, the dinosaur's texture is completely black yet flickers on and off. 

![Cartoon-Shade Issue](https://user-images.githubusercontent.com/69178113/89246245-8b413e00-d5bf-11ea-8602-841169c8bb5a.PNG)

The other sample is the fireflies sample. Here the words on the screen (such as ESC: Quit) constantly flicker on and off. 

![Fireflies Issue](https://user-images.githubusercontent.com/69178113/89246440-fab72d80-d5bf-11ea-94a8-cb691ab69110.PNG)

Interestingly, there doesn't seem to be an issue with the words in the cartoon shader, and there doesn't seem to be an issue with the textures in the fireflies sample.

## Environment
 This is on Windows 10 and again seems to only be a problem with the 5700XT.
The version of Panda is 1.10.6.

**Comment by Zoomao** [comment]
An interesting note is that according to multiple users, this isn't a problem on the AMD 20.4.2 version, but is a problem on every other version.

**Comment by Moguri** [comment]
I'm guessing there is no log output of interest (e.g., warnings from the graphics API)? Is this with Direct3D or OpenGL? In any case, we can at least try increasing notify levels to see if the their is some driver warning (doubtful, but I'm not sure where else to start with this):

```
gl-debug true
notify-level-glgsg debug
notify-level-gsg debug
```

You could also try `spam` instead of `debug`, but that might get hard to follow.

**Comment by rdb** [comment]
These issues tend to be hard to debug; the only way I can think of is just to use apitrace on a computer with the offending graphics card to poke around in the OpenGL state and make changes to see if they affect things, or if this is a shader bug (which this might well be), isolating the problem line by line in the shader until the culprit is found.

**Comment by MadaraZ** [comment]
I thought AMD would only work better OS Linux driver mesa than Windows (my graphics card rx 5500 xt nitro+)

@Zoomao if you are on windows 7 and higher, take the GTX/RTX video cards of the series or the Linux OS better take the Radeon RX 5500+series :)

Linux is based on GL and Windows DX these different renders, if Vulkan is okay there will be no problem.

**Comment by Moguri** [comment]
@Zoomao Can you confirm if this is still an issue with current drivers? I suspect this is a driver bug, and I am hoping AMD gets their OpenGL Windows drivers sorted out.

**Comment by tyteen4a03** [comment]
Can reproduce issue on Fireflies with 5700XT on Windows 10 (driver version 21.2.2).

**Comment by rdb** [comment]
Would it be possible for someone to lend me an affected card so that I can debug this issue?

**Comment by rdb** [comment]
A kind member of the community has loaned me an affected card, and I can reproduce the severe flickering in the Fireflies sample.  Will investigate.

**Comment by rdb** [comment]
So, it appears to be the case that using glColorPointer with GL_BGRA is buggy on these cards.  It does not appear to be a problem with glVertexAttribPointer.  Enabling the shader generator on affected nodes is enough to work around the problem.  Custom shaders may still exhibit the issue, because the shader generator uses special `ATTR#` semantics for vertex inputs, whereas custom shaders tend to use eg. `COLOR` semantics, which cause these inputs to be routed through `glColorPointer`.

As a workaround, I will disable the use of GL_BGRA vertex inputs if it detects that these cards are in use.  This does cause a minor performance regression for these cards, since geometry data has to be munged before it can be uploaded to the GPU.

The way to avoid this performance regression is to change Panda to never create vertex data as packed BGRA style.  However, this will cause an issue with DirectX 9, because it doesn't support anything other than BGRA vertex colors.  We can't automatically detect this, because assets may be loaded before the (fallback) renderer is loaded.

Maybe we need an additional configuration flag to switch Panda to OpenGL-style vertex colors, for people who know they do not need to use DirectX 9.

**Comment by rdb** [comment]
In 1.10.9, I have applied the workaround.  I have also added the `gl-support-vertex-array-bgra` config switch which can be set to `false` if the issue pops up on other cards.

I have also added `vertex-colors-prefer-packed` to master, which can be set to `false` for games that only use OpenGL and do not want to optimize vertex colors for DirectX 9 by default.  This should also avoid the problem, but without the (very minor) performance penalty.

**Comment by Omeryl** [comment]
Confirmed working here, thanks!

**Comment by rdb** [comment]
Good to hear!

I reported the bug to AMD in the meantime:
https://community.amd.com/t5/opengl-vulkan/rdna-opengl-bug-with-glcolorpointer-and-gl-bgra/m-p/459985

**Comment by Seltraeh** [comment]
So is the issue the colors being used or is it the outdated textures?
